sycl/doc/SharedLibraries.md

```diff
@@ -1,52 +1,84 @@
 library, then use `link` SYCL API to get a single program and launch kernels
 using it. But it is not user-friendly and it is very different from regular
 C/C++ workflow.
 
+Another possible scenario - use functions defined in pre-compiled device image
+provided by user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void foo();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { foo(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void foo() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option (like `-fsycl-add-targets`).
+
 The main purpose of this feature is to provide a mechanism which allows to
-provide *device* functions via shared libraries and works as close as possible
-to regular shared libraries.
-
-## Requrements:
-User's code is compiled into a shared library which consists of some host API,
-device code and device API (`SYCL_EXTERNAL` functions). The library is linked to
-a user's application which also contains some device code and performs
-computations using DPC++/SYCL.
+link device code dynamically at runtime.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
 For this combination the following statements must be true:
 
-- `SYCL_EXTERNAL` functions from library can be called (directly or indirectly)
-  from device code of the application.
-- Function pointers taken in application should work inside the library.
+- `SYCL_EXTERNAL` functions defined in dynamically linked code can be called
+  (directly or indirectly) from device code of the application.
+- Function pointers taken in application should work inside the dynamically
+  linked code.
 - Specific code changes are not required, i.e. the mechanism of linking works
   as close as possible to regular shared libraries.
 
 ## Design
-The overall idea is simple:
+The overall idea:
 
-- Each device image is supplied with an information about exported and imported
-  symbols using device image properties
-- DPC++ RT performs *device images collection* task by grouping all device
-  images required to execute a kernel based on the list of exports/imports
+- Each device image is supplied with a list of imported and exported symbol
+  names through device image properties mechanism
+- Before compiling a device image DPC++ RT will check if device image has a list
+  of imported symbols and if it has, then RT will search for device images which
+  define required symbols using lists of exported symbols.
```

Comment:
A note regarding "- Actual linking is performed by underlying backend (OpenCL/L0/etc.)" below.
Please specify what is the "underlying backend" in this case. There can be multiple co-existing in a SYCL RT, how the one is selected.

Reply:
I'm not sure that it is a part of current design. AFAIK backend is selected by user or by SYCL RT itself and this mechanism is independent from dynamic linking. I can use "target backend" or "chosen backend" here. Will this make sense?

Reply:
I guess this is a matter of defining the "underlying" term. Maybe a note like "Underlying backend is the backend used by the SYCL RT to perform JIT compilation of the program with symbols that need dynamic resolution". There can be multiple such requests in parallel, and different BEs may be used (e.g. opencl/CPU and level_zero/GPU).

Reply:
Done.